### PR TITLE
Fix #4493 and #4514

### LINKED ISSUE #4493: Failure to search for implicit type tag inside ~ of macro
Both
```scala
class Index[K]
object Index {
  inline def succ[K]: Unit = ~{
    implicit val k = '[K]
    '(new Index[K])
  }
}
```

```scala
class Index[K]
object Index {
  inline def succ[K]: Unit = ~{
    '(new Index[K])
  }
}
```

fail with 
```
Exception in thread "main" java.lang.AssertionError: assertion failed: unresolved symbols: type K when pickling tests/run/tasty-indexed-map/quoted_1.scala
	at scala.Predef$.assert(Predef.scala:219)
	at dotty.tools.dotc.core.tasty.TreePickler.pickle(TreePickler.scala:639)
	at dotty.tools.dotc.core.quoted.PickledQuotes$.pickle(PickledQuotes.scala:98)
	at dotty.tools.dotc.core.quoted.PickledQuotes$.pickleQuote(PickledQuotes.scala:31)
	at dotty.tools.dotc.transform.ReifyQuotes$Reifier.pickleAsTasty$1(ReifyQuotes.scala:401)
	at dotty.tools.dotc.transform.ReifyQuotes$Reifier.pickledQuote(ReifyQuotes.scala:404)
	at dotty.tools.dotc.transform.ReifyQuotes$Reifier.quotation(ReifyQuotes.scala:389)
	....
```

### PR DESCRIPTION
**Pull request by nicolasstucki**
Use captured type tags for macro type parameters. Before this change we just kept the type and failed when we tied to pickle it.

### CODE REVIEW

File: compiler/src/dotty/tools/dotc/transform/ReifyQuotes.scala [hunk 1]
```diff
@@ -144,6 +144,9 @@ class ReifyQuotes extends MacroTransformWithImplicits with InfoTransformer {
     /** We are in a `~(...)` context that is not shadowed by a nested `'(...)` */
     def inSplice = outer != null && !inQuote
 
+    /** We are not in a `~(...)` or a `'(...)` */
+    def isRoot = outer != null
```
Comment (biboudis) [inline compiler/src/dotty/tools/dotc/transform/ReifyQuotes.scala]: Hmm. I am confused. Isn't this root if the outer *is* `null` according to line 128 above? 

(also a typo in the comment of the description of the parameter `outer` )

Comment (nicolasstucki) [reply compiler/src/dotty/tools/dotc/transform/ReifyQuotes.scala]: Actually it is a typo. It should be `isNotRoot`

Comment (nicolasstucki) [reply compiler/src/dotty/tools/dotc/transform/ReifyQuotes.scala]: Or I will simplify it to an actual `isRoot` and add the not at call site.

Comment (biboudis) [reply compiler/src/dotty/tools/dotc/transform/ReifyQuotes.scala]: Good to go, then. Thanks!

**Pull request merged**